README.md
<h1 align="center">mimcvdf ⏲️</h1>

<p align="center"><i>Simple <a href="https://eprint.iacr.org/2018/601.pdf">Verifiable Delay Function</a> using <a href="https://eprint.iacr.org/2016/492.pdf">MiMC</a></i></p>

## Applications

This module was created for use in reducing spam in a similar manner to [HashCash](https://en.wikipedia.org/wiki/Hashcash). However, some potential uses for VDFs include blockchains and verifiable lotteries.


## Usage

```
from mimcvdf import vdf_create, vdf_verify


# Get a mimc hash of a byte sequence

vdf_create(byte_data, round_count) # Returns hex string


# Verify a mimc hash (must use same round count)

vdf_verify(same_bytes_data, vdf_create_result, rounds)

```

For high speed, you need GMP installed. Otherwise, the module will fall back to a Python implementation that is significantly slower.


## Security


This code has not been audited for security. It is not currently recommended for protecting against anything but denial of service, even though MiMC is able to do much more. If used, make sure it is not protecting anything critical.

In addition, since this is a Python implementation, attackers can be faster than typical users of an application by using faster language implementations or even FPGAs.



CHANGELOG.md
# Changelog

This project uses Semantic Versioning

## 1.2.0

+ Accepted C+gmp implementation of MiMC from @cartr
+ Added is_fast in mimcvdf to tell which implementation is being used


## 1.1.0

Added support for decimal output



tests/test_vdf.py
import sys
import os
sys.path.append('..')
import unittest
from time import time
from hashlib import sha3_256

import mimcvdf

class TestVDF(unittest.TestCase):

    def test_both(self):
        h = mimcvdf.vdf_create(b"test")
        self.assertTrue(mimcvdf.vdf_verify(b"test", h))

    def test_invalid_both(self):
        h = mimcvdf.vdf_create(b"test")
        self.assertFalse(mimcvdf.vdf_verify(b"invalid", h))

    def test_above_zero_rounds(self):
        self.assertRaises(AssertionError, mimcvdf.vdf_create, b"test", 0)
        self.assertRaises(AssertionError, mimcvdf.vdf_create, b"test", -1)
        self.assertRaises(AssertionError, mimcvdf.vdf_create, b"test", 1)
        self.assertRaises(AssertionError, mimcvdf.vdf_create, b"test", -10000)

    def test_dec_false(self):
        for i in range(3):
            h = mimcvdf.vdf_create(os.urandom(i), 2, dec=False)
            self.assertTrue(h.isalnum())
            self.assertTrue(len(h) == 63 or len(h) == 64)
            self.assertIs(type(h), str)

    def test_dec_true(self):
        for i in range(3):
            h = mimcvdf.vdf_create(os.urandom(i), 2, dec=True)
            self.assertIs(type(h), int)

    def test_dec_true_verify(self):
        h = mimcvdf.vdf_create(b"test", dec=True)
        self.assertTrue(mimcvdf.vdf_verify(b"test", h))

    def test_hash_starting_with_zero(self):
        s = b"test vector 1097"
        self.assertEqual(0, sha3_256(s).digest()[0])
        h = mimcvdf.vdf_create(s)
        self.assertTrue(mimcvdf.vdf_verify(s, h))


unittest.main()


tests/test_mimc.py
import sys
import os
sys.path.append('..')
import unittest
import time
from hashlib import sha3_256

import mimcvdf.mimc

class TestMimc(unittest.TestCase):

    def test_bytes(self):
        start = time.time()
        data = b"a" * 6000000
        h = sha3_256()
        h.update(data)
        data = h.digest()

        forw = mimcvdf.forward_mimc(data, 2000)

        rev = mimcvdf.reverse_mimc(forw, 2000)

        print(data.hex())
        print(forw.hex(), rev.hex())
        self.assertEqual(rev, data)

    def test_expected_data(self):
        h = sha3_256()
        h.update(b"test")
        data = h.digest()
        self.assertEqual(
            "66ea2a863bd103f2c7f190503cf8456198f31660069d4903afbd5f2e40a28695",
            mimcvdf.forward_mimc(data, 2000).hex()
        )

unittest.main()


mimcvdf/__init__.py
"""High level Verifiable Delay Function using keccak (sha3)."""
from hashlib import sha3_256
from typing import Union
import time
from statistics import mean
from math import ceil

from .mimc import forward_mimc, reverse_mimc, is_fast
"""
Kevin Froman 2020

This program is free software: you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation, either version 3 of the License, or
(at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program.  If not, see <https://www.gnu.org/licenses/>.
"""
DEFAULT_ROUNDS = 8000


def _sha3_256_hash(data: bytes) -> bytes:
    sha3 = sha3_256()
    sha3.update(data)
    return sha3.digest()


def vdf_create(data: bytes, rounds: int = DEFAULT_ROUNDS, dec=False) -> str:
    assert rounds > 1
    input_data: int = _sha3_256_hash(data)
    if dec:
        return int.from_bytes(reverse_mimc(input_data, rounds), "big")
    return reverse_mimc(input_data, rounds).hex()


def vdf_verify(
        data: bytes,
        test_hash: str,
        rounds: int = DEFAULT_ROUNDS) -> bool:
    """Verify data for test_hash generated by vdf_create."""
    assert rounds > 1
    should_match = _sha3_256_hash(data).lstrip(b'\0')
    if isinstance(test_hash, int):
        test_hash = test_hash.to_bytes((test_hash.bit_length() + 7) // 8, "big")
    else:
        try:
            test_hash = bytes.fromhex(test_hash)
        except ValueError:
            return False
    return forward_mimc(test_hash, rounds) == should_match


def profile_cpu_speed(seconds=1) -> float:
    n = 2
    start = time.time()
    done = False
    results = []
    try:
        for _ in range(20):
            done = False
            n = 2
            start = time.time()
            while not done:
                vdf_create(b't', n)
                if time.time() - start >= seconds:
                    break
                n += 1
            results.append(n)
    except KeyboardInterrupt:
        pass
    return ceil(mean(results))


if __name__ == "__main__":
    print("Calculate how may rounds are needed for X seconds (influenced by system processes): ")
    seconds = int(input("Seconds: "))
    print("Rounds:", profile_cpu_speed(seconds))


mimcvdf/mimc/fallback.py
"""Slow fallback implementation of MIMC written in pure Python."""
"""
This module adapted from https://github.com/OlegJakushkin/deepblockchains/blob/master/vdf/mimc/python/mimc.py by Sourabh Niyogi https://github.com/sourabhniyogi

This program is free software: you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation, either version 3 of the License, or
(at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program.  If not, see <https://www.gnu.org/licenses/>.
"""

is_fast = False

_modulus = 2**256 - 2**32 * 351 + 1
_little_fermat_expt = (_modulus*2-1)//3
_round_constants = [(i**7) ^ 42 for i in range(64)]


def forward_mimc(input_data: bytes, steps: int) -> bytes:
    inp = int.from_bytes(input_data, "big")
    for i in range(1,steps):
        inp = (inp**3 + _round_constants[i % len(_round_constants)]) % _modulus
    return inp.to_bytes((inp.bit_length() + 7) // 8, "big")


def reverse_mimc(input_data: bytes, steps: int) -> bytes:
    rtrace = int.from_bytes(input_data, "big")

    for i in range(steps - 1, 0, -1):
        rtrace = pow(rtrace-_round_constants[i%len(_round_constants)],
                     _little_fermat_expt, _modulus)
    return rtrace.to_bytes((rtrace.bit_length() + 7) // 8, "big")


mimcvdf/mimc/native.c
#define PY_SSIZE_T_CLEAN
#include <Python.h>

#include <stdbool.h>
#include <gmp.h>

/*
C Version by github.com/cartr
This module adapted from https://github.com/OlegJakushkin/deepblockchains/blob/master/vdf/mimc/python/mimc.py by Sourabh Niyogi https://github.com/sourabhniyogi

This program is free software: you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation, either version 3 of the License, or
(at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program.  If not, see <https://www.gnu.org/licenses/>.
*/

static mpz_t MODULUS;
static mpz_t LITTLE_FERMAT_EXPT;
#define ROUND_CONSTANTS_COUNT 64
static mpz_t ROUND_CONSTANTS[ROUND_CONSTANTS_COUNT];

static void
mimc_init_constants()
{
    mpz_t fortytwo;

    mpz_init(fortytwo);
    mpz_set_ui(fortytwo, 42);

    // Set MODULUS to hex(2**256 - 2**32 * 351 + 1)
    mpz_init(MODULUS);
    mpz_set_str(MODULUS,
                "ffffffffffffffffffffffffffffffff"
                "fffffffffffffffffffffea100000001",
                16);

    // Set LITTLE_FERMAT_EXPT to hex((MODULUS * 2 - 1) // 3)
    mpz_init(LITTLE_FERMAT_EXPT);
    mpz_set_str(LITTLE_FERMAT_EXPT,
                "aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa"
                "aaaaaaaaaaaaaaaaaaaaa9c0aaaaaaab",
                16);

    // Set all the constants to (i**7 ^ 42)
    for (unsigned long int i = 0; i < ROUND_CONSTANTS_COUNT; i++) {
       mpz_init(ROUND_CONSTANTS[i]);
       mpz_ui_pow_ui(ROUND_CONSTANTS[i], i, 7);
       mpz_xor(ROUND_CONSTANTS[i], ROUND_CONSTANTS[i], fortytwo);
    }

    mpz_clear(fortytwo);
}

static bool
unpack_args(PyObject *args, mpz_t input, unsigned int *steps)
{
    const char *data_bytes;
    Py_ssize_t count;

    if (!PyArg_ParseTuple(args, "y#I", &data_bytes, &count, steps)) {
        return false;
    }

    mpz_import(input, count, 1, 1, 0, 0, data_bytes);
    return true;
}


static PyObject *
convert_mpz_to_bytes(mpz_t op)
{
    char string[34];
    size_t size;

    mpz_export(string, &size, 1, 1, 0, 0, op);
    PyObject *result = PyBytes_FromStringAndSize(string, size);
    return result;
}

static PyObject *
forward_mimc(PyObject *_self, PyObject *args)
{
    mpz_t result;
    unsigned int steps;

    mpz_init(result);
    if (!unpack_args(args, result, &steps)) {
        mpz_clear(result);
        return NULL;
    }

    for (unsigned int i = 1; i < steps; ++i) {
        mpz_powm_ui(result, result, 3, MODULUS);
        mpz_add(result, result, ROUND_CONSTANTS[i % ROUND_CONSTANTS_COUNT]);
        if (mpz_cmp(result, MODULUS) >= 0) {
            mpz_sub(result, result, MODULUS);
        }
    }

    PyObject *result_obj = convert_mpz_to_bytes(result);
    mpz_clear(result);
    return result_obj;
}

static PyObject *
reverse_mimc(PyObject *_self, PyObject *args)
{
    mpz_t result;
    unsigned int steps;

    mpz_init(result);
    if (!unpack_args(args, result, &steps)) {
        mpz_clear(result);
        return NULL;
    }

    for (unsigned int i = steps - 1; i > 0; --i) {
        mpz_sub(result, result, ROUND_CONSTANTS[i % ROUND_CONSTANTS_COUNT]);
        mpz_powm(result, result, LITTLE_FERMAT_EXPT, MODULUS);
    }

    PyObject *result_obj = convert_mpz_to_bytes(result);
    mpz_clear(result);
    return result_obj;
}

static PyMethodDef Methods[] = {
    {"forward_mimc", forward_mimc, METH_VARARGS,
     "Run MIMC forward (the fast direction)"},
    {"reverse_mimc", reverse_mimc, METH_VARARGS,
     "Run MIMC in reverse (the slow direction)"},
    {NULL, NULL, 0, NULL}
};

static struct PyModuleDef moduledef = {
    PyModuleDef_HEAD_INIT,
    "native",
    "Fast native implementation of MIMC.",
    -1,
    Methods,
    NULL
};

PyMODINIT_FUNC
PyInit_native()
{
    mimc_init_constants();

    return PyModule_Create(&moduledef);
}
